File: src/main/java/no/jstien/bikeapi/tsdb/write/DatumBuilder.java
package no.jstien.bikeapi.tsdb.write;

import no.jstien.bikeapi.tsdb.Datum;

import java.util.ArrayList;
import java.util.List;

public class DatumBuilder {
    private String metricName;
    private List<String> tagKeys = new ArrayList<>();
    private TSDBWriter tsdbWriter;


    public DatumBuilder(String metricName, TSDBWriter tsdbWriter) {
        this.metricName = metricName;
        this.tsdbWriter = tsdbWriter;
    }

    public DatumBuilder addTagKey(String tagKey) {
        tagKeys.add(tagKey);
        return this;
    }

    public void addDatum(String ...tagValues) {
        addDatum(1L, tagValues);
    }

    public void addDatum(long value, String ...tagValues) {
        verifyTagValues(tagKeys, tagValues);

        Datum datum = new Datum(metricName, value);
        for (int i=0; i<tagKeys.size(); i++) {
            datum.addTag(tagKeys.get(i), tagValues[i]);
        }

        tsdbWriter.addDatum(datum);
    }


    private void verifyTagValues(List<String> keys, String[] values) {
        if (keys.size() != values.length) {
            StringBuilder sb = new StringBuilder();
            sb.append("Expected " + keys.size() + " tag values, got " + values.length + "! ");
            sb.append("Expectedd values for keys: ");
            keys.forEach(key -> sb.append(key + " "));
            throw new RuntimeException(sb.toString());
        }
    }
}

File: src/main/java/no/jstien/bikeapi/tsdb/write/DatumBuilder.kt
package no.jstien.bikeapi.tsdb.write

import java.util.ArrayList

class DatumBuilder(private val metricName: String, private val tsdbWriter: TSDBWriter) {
    private val tagKeys = ArrayList<String>()

    fun addTagKey(tagKey: String): DatumBuilder {
        tagKeys.add(tagKey)
        return this
    }

    fun addDatum(vararg tagValues: String) {
        addDatum(1L, *tagValues)
    }

    fun addDatum(value: Long, vararg tagValues: String) {
        verifyTagValues(tagKeys, arrayOf(*tagValues))

        val datum = Datum(metricName, value)
        for (i in tagKeys.indices) {
            datum.addTag(tagKeys[i], tagValues[i])
        }

        tsdbWriter.addDatum(datum)
    }


    private fun verifyTagValues(keys: List<String>, values: Array<String>) {
        if (keys.size != values.size) {
            val sb = StringBuilder()
            sb.append("Expected " + keys.size + " tag values, got " + values.size + "! ")
            sb.append("Expectedd values for keys: ")
            keys.forEach { key -> sb.append("$key ") }
            throw RuntimeException(sb.toString())
        }
    }
}
